src/exp_and_gram.jl

```diff
@@ -34,15 +34,58 @@ function exp_and_gram(
     return exp_and_gram!(similar(A), similar(A), copy(A), copy(B), t, method)
 end
 
+function alloc_mem(A, B, method::ExpAndGram{T,q}) where {T,q}
+    n, m = size(B)
+    ncoeffhalf = div(q + 1, 2)
+    return (
+        U = similar(A),
+        pre_array = similar(A, 2n, n),
+        tmp = similar(A),
+        _A = similar(A),
+        _B = similar(B),
+        P = similar(A),
+        A2 = similar(A),
+        odd = similar(A),
+        even = similar(A),
+        tmpA1 = similar(A),
+        tmpA2 = similar(A),
+        tmpA3 = similar(A),
+        L = zeros(eltype(A), n, m * (q + 1)),
+        Loddtmp = similar(B, n, m*ncoeffhalf),
+    )
+end
+function alloc_mem(A, B, method::ExpAndGram{T,13}) where {T}
+    q = 13
+    n, m = size(B)
+    return (
+        _A = similar(A),
+        _B = similar(B),
+        A2 = similar(A),
+        A4 = similar(A),
+        A6 = similar(A),
+        tmpA1 = similar(A),
+        tmpA2 = similar(A),
+        tmpA3 = similar(A),
+        L = similar(A, n, m * (q + 1)),
+        tmpB2 = similar(B),
+        A2B = similar(B),
+        A4B = similar(B),
+        A6B = similar(B),
+        U = similar(A),
+        pre_array = similar(A, 2n, n),
+        tmp = similar(A),
+    )
+end
+
 function exp_and_gram!(
     eA::AbstractMatrix{T},
     U::AbstractMatrix{T},
     A::AbstractMatrix{T},
     B::AbstractMatrix{T},
     method::AbstractExpAndGramAlgorithm,
-    cache = nothing,
+    cache = alloc_mem(A, B, method),
 ) where {T<:Number}
-    Φ, U = exp_and_gram_chol!(eA, U, A, B, method)
+    Φ, U = exp_and_gram_chol!(eA, U, A, B, method, cache)
     G = U' * U
```

Comment:
I think we should be able to do ```G = U' * U``` in-place here by using the cache? Also as ```G``` so its contents must be written into```U```.

Reply:
But cache could be nothing in the case of the adaptive method :/

Reply:
So ```G = _make_gramian!(U, cache)``` or so?

Reply:
Would this be fine?
```julia
    G = isnothing(cache) ? copy(U) : cache.U
    mul!(G, U', U)
```

Reply:
Yes!

Reply:
done